gemma/weights.h

```diff
@@ -0,0 +1,288 @@
+    type == LayerAttentionType::kGemma) {                               \
+    CALL_LAYER_FUNC ## N("attn_ob", attention_output_biases);           \
+  }
+
+template <typename T, typename TConfig, class Func>
+void ForEachTensor1(Func& func, const Weights<T, TConfig>& weights1) {
+  CALL_TOP_FUNC1("embedding", embedder_input_embedding);
+  CALL_TOP_FUNC1("final_norm", final_norm_scale);
+  char name_buf[16];
+  for (int layer_idx = 0; layer_idx < TConfig::kLayers; ++layer_idx) {
+    auto type = TConfig::kLayerConfig[layer_idx];
+    const size_t idx = static_cast<size_t>(layer_idx);
+    const LayerF<TConfig>& layer1 = *weights1.GetLayer(idx);
+    CALL_ALL_LAYER_FUNC(1)
+  }
+}
+
+template <typename T, typename TConfig, class Func>
+void ForEachTensor1(Func& func, Weights<T, TConfig>& weights1) {
+  CALL_TOP_FUNC1("embedding", embedder_input_embedding);
+  CALL_TOP_FUNC1("final_norm", final_norm_scale);
+  char name_buf[16];
+  for (int layer_idx = 0; layer_idx < TConfig::kLayers; ++layer_idx) {
+    auto type = TConfig::kLayerConfig[layer_idx];
+    const size_t idx = static_cast<size_t>(layer_idx);
+    LayerF<TConfig>& layer1 = *weights1.GetLayer(idx);
+    CALL_ALL_LAYER_FUNC(1)
+  }
+}
+
+template <typename T, typename TConfig, class Func>
+void ForEachTensor2(Func& func, const Weights<T, TConfig>& weights1,
+                    Weights<T, TConfig>& weights2) {
+  CALL_TOP_FUNC2("embedding", embedder_input_embedding);
+  CALL_TOP_FUNC2("final_norm", final_norm_scale);
+  char name_buf[16];
+  for (int layer_idx = 0; layer_idx < TConfig::kLayers; ++layer_idx) {
+    auto type = TConfig::kLayerConfig[layer_idx];
+    const size_t idx = static_cast<size_t>(layer_idx);
+    const LayerF<TConfig>& layer1 = *weights1.GetLayer(idx);
+    LayerF<TConfig>& layer2 = *weights2.GetLayer(idx);
+    CALL_ALL_LAYER_FUNC(2)
+  }
+}
+
+#undef CALL_TOP_FUNC1
+#undef CALL_TOP_FUNC2
+#undef CALL_TOP_FUNC3
+#undef CALL_TOP_FUNC4
+#undef CALL_LAYER_FUNC1
+#undef CALL_LAYER_FUNC2
+#undef CALL_LAYER_FUNC3
+#undef CALL_LAYER_FUNC4
+#undef CALL_ALL_LAYER_FUNC
+
+template<typename T, typename TConfig>
+void ZeroInit(Weights<T, TConfig>& w) {
+  memset(&w.embedder_input_embedding, 0, sizeof(w.embedder_input_embedding));
+  memset(&w.final_norm_scale, 0, sizeof(w.final_norm_scale));
+  for (int i = 0; i < TConfig::kLayers; ++i) {
+    memset(w.GetLayer(i), 0, sizeof(*w.GetLayer(i)));
+  }
+}
+
+template<typename T, typename TConfig>
+void Copy(Weights<T, TConfig>& dst, const Weights<T, TConfig>& src) {
+  memcpy(&dst.embedder_input_embedding, &src.embedder_input_embedding,
+         sizeof(src.embedder_input_embedding));
+  memcpy(&dst.final_norm_scale, &src.final_norm_scale,
+         sizeof(src.final_norm_scale));
+  for (int i = 0; i < TConfig::kLayers; ++i) {
+    memcpy(dst.GetLayer(i), src.GetLayer(i), sizeof(*dst.GetLayer(i)));
+  }
+}
+
+template<typename T, typename TConfig>
+class WeightsWrapper {
```

Comment:
Comment on what this is for? Something like "owns weights and undoes the type erasure of AllocateWeights"?

Reply:
Done.